WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/FilterableCategoriesViewController.swift

```diff
@@ -0,0 +1,237 @@
+import UIKit
+import Gridicons
+import Gutenberg
+
+class FilterableCategoriesViewController: CollapsableHeaderViewController {
+    let tableView: UITableView
+    internal var selectedItem: IndexPath? = nil {
+        didSet {
+            if !(oldValue != nil && selectedItem != nil) {
+                itemSelectionChanged(selectedItem != nil)
+            }
+        }
+    }
+    private let filterBar: CollapsableHeaderFilterBar
+
+    /* This should be overriden by the subclass to provide a conforming collection of categories */
+    internal var categorySections: [CategorySection] { get { [] } }
```

Comment:
I'm not thrilled with this pattern (as it offers no compiler signaling to the subclass). For example, this would have been nicer as protocol getter (along with the suggested snippet the subclass implementation would get for free). But in my explorations, I didn't find a valid inheritence heirarchy to accomplish that. This seemed acceptable to me, all things considered, but open to suggestions if you have any :smiley:

Reply:
Thanks for providing some ideas here Chip!

> 1. Don't make this a computed property and have it be the base for `sections` in the subclasses. This wouldn't provide the compiler hint but would reduce the duplicated variables in the subclasses.

I had initially tried this, but it won't work because generics are not covariant:

`Property 'sections' with type '[GutenbergLayoutSection]' cannot override a property with type '[CategorySection]'`

so, I introduced a computed property to alias the sections subtypes and perform the implicit upcasting.

> 2. We could have this be:
> ```
> private let dataSource: FilterableCategoriesDataSource
> ```
> We can then pass the dataSource in through the `init` and that dataSource could be something like:
> 
> ```
> protocol FilterableCategoriesDataSource {
>      var categorySections: [CategorySection] { get }
> }
> ```
> This provides the compiler hint and in the terms of the subclasses, `SiteDesigns` and `GutenbergLayoutSection` could conform to `FilterableCategoriesDataSource` 

I began investigating this approach, and I agree that this one seems the most promising, but it isn't as straightforward as it may at first seem. `GutenbergLayoutSection` in it's current form represents a _single_ section, and conformance to `FilterableCategoriesDataSource` would require it to _have a_ collection of `CategorySections`.

I also investigated making `GutenbergLayoutPickerViewController` conform to `FilterableCategoriesDataSource` (seemingly more appropriate regarding the plural _has a_  relationship with the `sections` array),__ and a different problem arises: it must pass `self` as the `FilterableCategoriesDataSource` when calling its `super.init`, but it cannot do so from within its own `init`, since `self` is not yet defined. We could make it `var dataSource` (instead of `let`) and assign in subclass `init` after the call to `super.init`, but that feels awkward: `self.dataSource = self`.

Generally, the collection must be variable since we assign asyncronously via remote callbacks, and I believe this restricts us from binding at init, and as a consequence, I think we forfeit that kind of compiler signalling. E.g. in the latter approach, there is nothing stopping someone from omiting the `dataSource` assignment while subclassing `FilterableCategoriesViewController` in `init`.

It's possible that I'm missing something in your suggestion, so if what I tried is not what you had in mind, please elaborate further.

> 3. The getter could provide a fatal error at runtime like `fatalError("This should be overridden by the subclass to provide a conforming collection of categories")`. This would force the implementation in the subclass. 

Would it make sense to include this behind a debug flag?

Reply:
> Would it make sense to include this behind a debug flag?

We could although for this use case it would closely align to our other [instances](https://github.com/wordpress-mobile/WordPress-iOS/search?q=fatalError) where this is would be a developer error and wouldn't be executed in `develop` unless a change to the screen was made and not tested. So I don't think a dev flag would be necessary.

Reply:
I've pushed an update to add the message as a fatalError in the base class.